src/network.js

```diff
@@ -0,0 +1,181 @@
+
+    setImmediate(() => callback())
+  }
+
+  // Handles both types of bitswap messgages
+  _onConnection (protocol, conn) {
+    if (!this._running) { return }
+    log('incomming new bitswap connection: %s', protocol)
+
+    pull(
+      conn,
+      lp.decode(),
+      pull.asyncMap((data, cb) => Message.deserialize(data, cb)),
+      pull.asyncMap((msg, cb) => {
+        conn.getPeerInfo((err, peerInfo) => {
+          if (err) { return cb(err) }
+
+          // log('data from', peerInfo.id.toB58String())
+          this.bitswap._receiveMessage(peerInfo.id, msg, cb)
+        })
+      }),
+      pull.onEnd((err) => {
+        log('ending connection')
+        if (err) { return this.bitswap._receiveError(err) }
+      })
+    )
+  }
+
+  _onPeerConnect (peerInfo) {
+    if (!this._running) { return }
+
+    this.bitswap._onPeerConnected(peerInfo.id)
+  }
+
+  _onPeerDisconnect (peerInfo) {
+    if (!this._running) { return }
+
+    this.bitswap._onPeerDisconnected(peerInfo.id)
+  }
+
+  findProviders (cid, maxProviders, callback) {
+    // TODO
+    // consider if we want to trickleDown maxProviders, currently this is
+    // not an exposed option:
+    // https://github.com/libp2p/js-libp2p-kad-dht/blob/master/src/index.js#L416
+    this.libp2p.contentRouting.findProviders(cid, CONSTANTS.providerRequestTimeout, callback)
+  }
+
+  findAndConnect (cid, maxProviders, callback) {
+    waterfall([
+      (cb) => this.findProviders(cid, maxProviders, cb),
+      (provs, cb) => each(provs, (p, cb) => this.connectTo(p, cb))
+    ], callback)
+  }
+
+  provide (cid, callback) {
+    this.libp2p.contentRouting.provide(cid, callback)
+  }
+
+  // Connect to the given peer
+  // Send the given msg (instance of Message) to the given peer
+  sendMessage (peer, msg, callback) {
+    if (!this._running) { return callback(new Error(`network isn't running`)) }
+
+    const stringId = peer.toB58String() ? peer.toB58String() : peer.id.toB58String()
+    log('sendMessage to %s', stringId, msg)
+
+    this._dialPeer(peer, (err, conn, protocol) => {
+      if (err) { return callback(err) }
+
+      let serialized
+      switch (protocol) {
+        case BITSWAP100: serialized = msg.serializeToBitswap100(); break
+        case BITSWAP110: serialized = msg.serializeToBitswap110(); break
+        default: return callback(new Error('Unkown protocol: ' + protocol))
+      }
+      // TODO: why doesn't the error get propageted back??
+      writeMessage(conn, serialized, (err) => {
+        if (err) { log(err) }
+      })
+      callback()
+    })
+  }
+
+  // TODO: if this method is just to call libp2p.dial, better call libp2p dial
+  connectTo (peer, callback) {
+    if (!this._running) { return callback(new Error(`network isn't running`)) }
+
+    this.libp2p.dial(peer, callback)
+  }
+
+  // Dial to the peer and try to use the most recent Bitswap
+  _dialPeer (peer, callback) {
+   // Attempt Bitswap 1.1.0
+    this.libp2p.dial(peer, BITSWAP110, (err, conn) => {
```

Comment:
Removing the try/catch block here showed that our mocks were actually not correct. The try catch around these calls was making so that any error that would happen after a dial would end up here to be dealt from bitswap possibly causing a lot of other erros.